packages/block-library/src/template-part/edit/index.js

```diff
@@ -27,44 +27,56 @@ export default function TemplatePartEdit( {
 	setAttributes,
 	clientId,
 } ) {
-	const initialPostId = useRef( _postId );
 	const initialSlug = useRef( slug );
 	const initialTheme = useRef( theme );
 
 	// Resolve the post ID if not set, and load its post.
 	const postId = useTemplatePartPost( _postId, slug, theme );
 
-	// Set the post ID, once found, so that edits persist,
-	// but wait until the third inner blocks change,
-	// because the first 2 are just the template part
-	// content loading.
-	const { innerBlocks } = useSelect(
+	// Set the postId attribute if it did not exist,
+	// but wait until the inner blocks have loaded to allow
+	// new edits to trigger this.
+	const { innerBlocks, expectedContent } = useSelect(
 		( select ) => {
 			const { getBlocks } = select( 'core/block-editor' );
+			const entityRecord = select( 'core' ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				postId
+			);
+
 			return {
 				innerBlocks: getBlocks( clientId ),
+				expectedContent: entityRecord?.content.raw,
 			};
 		},
-		[ clientId ]
+		[ clientId, postId ]
 	);
 	const { editEntityRecord } = useDispatch( 'core' );
-	const blockChanges = useRef( 0 );
+	const initialContent = useRef();
 	useEffect( () => {
-		if ( blockChanges.current < 4 ) blockChanges.current++;
+		if ( postId === null || postId === undefined ) {
+			return;
+		}
+
+		if ( ! initialContent.current ) {
+			const innerContent = serialize( innerBlocks );
```

Comment:
`serialize( innerBlocks )` is used in both conditionals, so I guess we can move this out of this `if`, and use the same `innerContent` variable in both.
It would hopefully simplify, even if just slightly, the readability of the hook.

Reply:
I like to keep the `serialize` in the conditions because it prevents running the serialization unnecessarily.  `initialContent.current` is only falsy for a short amount of time, limiting how often this `serialize` runs.  Similarly in the if statement below, the `serialize` only happens if the postId block attribute has not yet been set.

If we move the `serialize` out of these if statements and to another level above, then it would run unnecessarily even after the hook has taken care of all that it needs to.

Reply:
Although, I could probably restructure the logic in a way to get the best of both worlds. 🤔

Reply:
👍  - refactored the logic slightly in 5018d6b so we can do this.